Tested the negative tests before making the fix:

```
akhil-mac@Akhils-MacBook-Air gouroboros % 
akhil-mac@Akhils-MacBook-Air gouroboros % go test ./cbor/... -v
--- PASS: TestSetTypeWithoutTag (0.00s)
=== RUN   TestSetTypeMarshalCBOR
--- PASS: TestSetTypeMarshalCBOR (0.00s)
=== RUN   TestSetTypeMarshalCBORWithoutTag
--- PASS: TestSetTypeMarshalCBORWithoutTag (0.00s)
=== RUN   TestValueDecode
    value_test.go:174: failed to decode CBOR data: unsupported numerator type for cbor.Rat: big.Int
--- FAIL: TestValueDecode (0.00s)
=== RUN   TestValueMarshalJSON
    value_test.go:223: CBOR did not marshal to expected JSON
          got:    {"cbor":"d81e822002","json":{"list":[{"int":1},{"int":2}]}}
          wanted: {"cbor":"d81e822002","json":{"list":[{"int":-1},{"int":2}]}}
--- FAIL: TestValueMarshalJSON (0.00s)
=== RUN   TestLazyValueDecode
    value_test.go:265: failed to decode CBOR data: unsupported numerator type for cbor.Rat: big.Int
--- FAIL: TestLazyValueDecode (0.00s)
=== RUN   TestLazyValueMarshalJSON
    value_test.go:313: CBOR did not marshal to expected JSON
          got:    {"cbor":"d81e822002","json":{"list":[{"int":1},{"int":2}]}}
          wanted: {"cbor":"d81e822002","json":{"list":[{"int":-1},{"int":2}]}}
--- FAIL: TestLazyValueMarshalJSON (0.00s)
=== RUN   TestLazyValueMarshalJSONEmptyCborData
=== RUN   TestLazyValueMarshalJSONEmptyCborData/zero_value
=== RUN   TestLazyValueMarshalJSONEmptyCborData/empty_cbor_after_unmarshal
--- PASS: TestLazyValueMarshalJSONEmptyCborData (0.00s)
    --- PASS: TestLazyValueMarshalJSONEmptyCborData/zero_value (0.00s)
    --- PASS: TestLazyValueMarshalJSONEmptyCborData/empty_cbor_after_unmarshal (0.00s)
FAIL
FAIL    github.com/blinklabs-io/gouroboros/cbor 0.278s
FAIL
akhil-mac@Akhils-MacBook-Air gouroboros % 

```

--- a/cbor/tags.go
+++ b/cbor/tags.go
@@ -113,6 +113,8 @@ func (r *Rat) UnmarshalCBOR(cborData []byte) error {
 		tmpNum.SetInt64(v)
 	case uint64:
 		tmpNum.SetUint64(v)
+	case big.Int:
+		tmpNum.Set(&v)
 	default:
 		return fmt.Errorf("unsupported numerator type for cbor.Rat: %T", v)
 	}
@@ -123,6 +125,8 @@ func (r *Rat) UnmarshalCBOR(cborData []byte) error {
 		tmpDenom.SetInt64(v)
 	case uint64:
 		tmpDenom.SetUint64(v)
+	case big.Int:
+		tmpDenom.Set(&v)
 	default:
 		return fmt.Errorf("unsupported denominator type for cbor.Rat: %T", v)
 	}

--- a/cbor/value.go
+++ b/cbor/value.go
@@ -203,8 +203,8 @@ func generateAstJson(obj any) ([]byte, error) {
 	case Rat:
 		return generateAstJson(
 			[]any{
-				v.Num().Uint64(),
-				v.Denom().Uint64(),
+				v.Num(),
+				v.Denom(),
 			},
 		)
 	case int, uint, uint64, int64:

--- a/cbor/value_test.go
+++ b/cbor/value_test.go
@@ -93,7 +93,7 @@ var testDefs = []struct {
 		},
 		expectedAstJson: `{"list":[{"int":3},{"int":1000}]}`,
 	},
-	// 30([-1, 2])
+	// 30([-1, 2]) -- negative rational
 	{
 		cborHex: "d81e822002",
 		expectedObject: cbor.Rat{
@@ -101,18 +101,18 @@ var testDefs = []struct {
 		},
 		expectedAstJson: `{"list":[{"int":-1},{"int":2}]}`,
 	},
-	// 30([18446744073709551616, 2])
+	// 30([18446744073709551617, 1]) -- numerator = 2^64+1 greater than MaxUint64
 	{
-		cborHex: "d81e82c24901000000000000000002",
+		cborHex: "d81e82c24901000000000000000101",
 		expectedObject: cbor.Rat{
 			Rat: new(big.Rat).SetFrac(
-				new(big.Int).SetBytes(test.DecodeHexString("010000000000000000")),
-				big.NewInt(2),
+				new(big.Int).SetBytes(test.DecodeHexString("010000000000000001")),
+				big.NewInt(1),
 			),
 		},
-		expectedAstJson: `{"list":[{"int":18446744073709551616},{"int":2}]}`,
+		expectedAstJson: `{"list":[{"int":18446744073709551617},{"int":1}]}`,
 	},
-	// 30([1, 0])
+	// 30([1, 0]) -- zero denominator
 	{
 		cborHex:             "d81e820100",
 		expectedObject:      nil,

--- a/cbor/value_test.go
+++ b/cbor/value_test.go
@@ -93,6 +93,31 @@ var testDefs = []struct {
 		},
 		expectedAstJson: `{"list":[{"int":3},{"int":1000}]}`,
 	},
+	// 30([-1, 2])
+	{
+		cborHex: "d81e822002",
+		expectedObject: cbor.Rat{
+			Rat: big.NewRat(-1, 2),
+		},
+		expectedAstJson: `{"list":[{"int":-1},{"int":2}]}`,
+	},
+	// 30([18446744073709551616, 2])
+	{
+		cborHex: "d81e82c24901000000000000000002",
+		expectedObject: cbor.Rat{
+			Rat: new(big.Rat).SetFrac(
+				new(big.Int).SetBytes(test.DecodeHexString("010000000000000000")),
+				big.NewInt(2),
+			),
+		},
+		expectedAstJson: `{"list":[{"int":18446744073709551616},{"int":2}]}`,
+	},
+	// 30([1, 0])
+	{
+		cborHex:             "d81e820100",
+		expectedObject:      nil,
+		expectedDecodeError: fmt.Errorf("invalid cbor.Rat: denominator cannot be zero"),
+	},
 	// 258([1, 2, 3])
 	{
 		cborHex: "d9010283010203",

